File: dodam-system-domain/dodam-domain-rds/src/main/java/b1nd/dodam/domain/rds/nightstudy/service/NightStudyProjectService.java
package b1nd.dodam.domain.rds.nightstudy.service;

import b1nd.dodam.domain.rds.member.entity.Student;
import b1nd.dodam.domain.rds.nightstudy.entity.NightStudyProject;
import b1nd.dodam.domain.rds.nightstudy.exception.NightStudyNotFoundException;
import b1nd.dodam.domain.rds.nightstudy.repository.NightStudyProjectRepository;
import b1nd.dodam.domain.rds.support.enumeration.ApprovalStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class NightStudyProjectService {

    private final NightStudyProjectRepository repository;

    public NightStudyProject save(NightStudyProject nightStudyProject) {
        return repository.save(nightStudyProject);
    }

    public void delete(NightStudyProject nightStudyProject) {
        repository.delete(nightStudyProject);
    }

    public NightStudyProject getBy(Long id) {
        return repository.findById(id)
                .orElseThrow(NightStudyNotFoundException::new);
    }

    public List<NightStudyProject> getPendingProjects(LocalDate date) {
        return repository.findByStatusAndEndAtGreaterThanEqual(ApprovalStatus.PENDING, date);
    }

    public List<NightStudyProject> getAllowedProjects(LocalDate date) {
        return repository.findByStatusAndStartAtLessThanEqualAndEndAtGreaterThanEqual(ApprovalStatus.ALLOWED, date, date);
    }

    public List<NightStudyProject> getAllByDateRange(LocalDate date) {
    return repository.findByStartAtLessThanEqualAndEndAtGreaterThanEqual(date, date);
    }

    public List<NightStudyProject> getMyProjects(Student leader, LocalDate date) {
        return repository.findByLeaderAndStartAtLessThanEqualAndEndAtGreaterThanEqual(leader, date, date);
    }

    public List<NightStudyProject> getAllRoomsWithProjects(LocalDate date) {
       return repository.findByStatusAndStartAtLessThanEqualAndEndAtGreaterThanEqualOrderByRoom(ApprovalStatus.ALLOWED, date, date);
    }
}

File: services/service-night-study/src/main/kotlin/com/b1nd/dodamdodam/nightstudy/domain/project/service/NightStudyProjectService.kt
package com.b1nd.dodamdodam.nightstudy.domain.project.service

import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.enumeration.NightStudyStatus
import com.b1nd.dodamdodam.nightstudy.domain.project.entity.NightStudyProjectEntity
import com.b1nd.dodamdodam.nightstudy.domain.project.exception.NightStudyProjectNotFoundException
import com.b1nd.dodamdodam.nightstudy.domain.project.repository.NightStudyProjectRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.UUID

@Service
class NightStudyProjectService(
    private val nightStudyProjectRepository: NightStudyProjectRepository
) {

    fun save(project: NightStudyProjectEntity): NightStudyProjectEntity =
        nightStudyProjectRepository.save(project)

    fun getById(id: Long): NightStudyProjectEntity =
        nightStudyProjectRepository.findById(id).orElseThrow { NightStudyProjectNotFoundException() }

    fun getByUserId(userId: UUID): List<NightStudyProjectEntity> =
        nightStudyProjectRepository.findAllByUserId(userId)

    fun getAllowed(date: LocalDate): List<NightStudyProjectEntity> =
        nightStudyProjectRepository.findAllByStatusAndStartAtLessThanEqualAndEndAtGreaterThanEqual(
            NightStudyStatus.ALLOWED, date, date
        )

    fun getPending(): List<NightStudyProjectEntity> =
        nightStudyProjectRepository.findAllByStatus(NightStudyStatus.PENDING)

    fun delete(project: NightStudyProjectEntity) =
        nightStudyProjectRepository.delete(project)
}
